src/redux/sagas/utils/editStreamingPaymentMulticall.ts

```diff
@@ -0,0 +1,196 @@
+      streamingPaymentsAddress,
+    );
+
+  const amountInWei = amount
+    ? (moveDecimal(amount, tokenDecimals) as string)
+    : streamingPayment.amount;
+
+  const realStartTimestamp = startTimestamp || streamingPayment.startTime;
+  const realInterval = interval || Number(streamingPayment.interval);
+  const realEndCondition =
+    endCondition || streamingPayment.metadata?.endCondition;
+
+  if (!realEndCondition) {
+    throw new Error('End condition undefined and cannot be found in metadata');
+  }
+
+  const limitInWei = limitAmount
+    ? (moveDecimal(limitAmount, tokenDecimals) as string)
+    : getStreamingPaymentLimit({ streamingPayment });
+
+  const realEndTimestamp = getEndTimeByEndCondition({
+    endCondition: realEndCondition,
+    startTimestamp: realStartTimestamp,
+    interval: realInterval,
+    amountInWei,
+    limitInWei,
+    endTimestamp: endTimestamp || streamingPayment.endTime,
+  });
+
+  const multicallData: string[] = [];
+
+  const hasAmountChanged = amountInWei !== streamingPayment.amount;
+  const hasIntervalChanged = realInterval !== Number(streamingPayment.interval);
+
+  if (hasAmountChanged || hasIntervalChanged) {
+    multicallData.push(
+      streamingPaymentsClient.interface.encodeFunctionData('setTokenAmount', [
+        fundingPermissionDomainId,
+        fundingChildSkillIndex,
+        extensionPermissionDomainId,
+        extensionChildSkillIndex,
+        extensionChildSkillIndex,
+        extensionChildSkillIndex,
+        streamingPayment.nativeId,
+        amountInWei,
+        realInterval,
+      ]),
+    );
+  }
+
+  const hasStartTimeChanged = streamingPayment.startTime !== realStartTimestamp;
+  const hasEndTimeChanged = !realEndTimestamp.eq(streamingPayment.endTime);
+
+  const pushSetStartTime = (time: string) => {
+    multicallData.push(
+      streamingPaymentsClient.interface.encodeFunctionData('setStartTime', [
+        adminPermissionDomainId,
+        adminChildSkillIndex,
+        streamingPayment.nativeId,
+        time,
+      ]),
+    );
+  };
+
+  const pushSetEndTime = (time: BigNumber) => {
+    multicallData.push(
+      streamingPaymentsClient.interface.encodeFunctionData('setEndTime', [
+        adminPermissionDomainId,
+        adminChildSkillIndex,
+        streamingPayment.nativeId,
+        time,
+      ]),
+    );
+  };
+
+  // @NOTE: The multicall will fail if we set the new start time to be before the current end time
+  // Or the new end time to be before the current start time
+  // If both are true, then first set the endTime to the maximum possible end time
+  if (
+    realStartTimestamp > streamingPayment.endTime &&
+    realEndTimestamp.lt(streamingPayment.startTime)
+  ) {
+    if (hasEndTimeChanged) {
+      pushSetEndTime(TIMESTAMP_IN_FUTURE);
+    }
+
+    if (hasStartTimeChanged) {
+      pushSetStartTime(realStartTimestamp);
+    }
+
+    if (hasEndTimeChanged) {
+      pushSetEndTime(realEndTimestamp);
+    }
```

Comment:
Is it intentional to call `pushSetEndTime()` twice in this scope due to some race condition or something? Because if that's the case, was thinking maybe we can get away with initialising some var to undefined then finally call `pushSetEndTime` at the end to save us multiple calls, so like:

```js
const endTime = undefined;
const startTime = undefined;

if (hasEndTimeChanged) {
  endTime = TIMESTAMP_IN_FUTURE;
}

if (hasStartTimeChanged) {
  startTime = realStartTimestamp;
}

if (hasEndTimeChanged) {
  endTime = realEndTimestamp;
}

// Then finally
pushSetEndTime(endTime);
pushSetStartTime(startTime);
```

But I dunno, maybe there's something more to this than meets the eye 😄

Reply:
It is intentional here and we want to call `setEndTime` twice as part of the same multicall.

The contract function doesn't allow for a new startTime that is after the current endTime, and also doesn't allow a new endTime that is before the current startTime. If just one or the other is true then we can just change the order in which we call `setStartTime` or `setEndTime`.

This section is to account for an edge case where both of these statements are true. To get around this, we set the endTime to the max value first, so that we can set the new startTime (which is after the current endTime) then finally set the actual endTime.

(Although now that I write this out, I can't think of a situation where it would be possible for both of these to be true at the same time...)